drivers/adc/adc_max32.c

```diff
@@ -150,6 +209,196 @@ static int adc_max32_read(const struct device *dev, const struct adc_sequence *s
+	[MAX32_12B_MODE] = 12,
+};
+
+uint8_t adc_max32_convert_vref_to_shift(uint16_t vref_mv)
+{
+	uint8_t count = 1;
+
+	while (1) {
+		vref_mv /= 2;
+		if (vref_mv) {
+			count++;
+		} else {
+			break;
+		}
+	}
+	return count;
+}
+
+static inline int adc_max32_convert_q31(q31_t *out, const uint8_t *buff,
+			enum max32_qscale_modes mode, uint8_t diff_mode,
+			uint16_t vref_mv, uint8_t adc_shift)
+{
+	int32_t data_in = 0;
+	uint32_t scale = BIT(adc_max32_resolution[mode]);
+
+	/* No Differential mode */
+	if (diff_mode) {
+		return -EINVAL;
+	}
+
+	uint32_t sensitivity = (vref_mv * (scale - 1)) / scale
+			 * 1000 / scale; /* uV / LSB */
+
+	if (mode == MAX32_12B_MODE) {
+		data_in = (buff[1] << 8) | buff[0];
+		if (diff_mode && (data_in & (BIT(adc_max32_resolution[mode] - 1)))) {
+			data_in |= ~BIT_MASK(adc_max32_resolution[mode]);
+		}
+	} else {
+		data_in = sys_get_be16(buff);
+		break;
+	}
+
+	*out =  BIT(31 - adc_shift) * sensitivity / 1000000 * data_in;
+	return 0;
+}
+
+static int adc_max32_decoder_get_frame_count(const uint8_t *buffer, uint32_t channel,
+			   uint16_t *frame_count)
+{
+	const struct adc_max32_fifo_data *data = (const struct adc_max32_fifo_data *)buffer;
+
+	*frame_count = data->fifo_byte_count/ADC_MAX32_SAMPLE_SIZE;
+
+	return 0;
+}
+
+static int adc_max32_decoder_decode(const uint8_t *buffer, uint32_t channel, uint32_t *fit,
+			  uint16_t max_count, void *data_out)
+{
+	const struct adc_max32_fifo_data *enc_data = (const struct adc_max32_fifo_data *)buffer;
+	uint8_t *out = (uint8_t *) data_out;
+
+	const uint8_t *buffer_end =
+		buffer + sizeof(struct adc_max32_fifo_data) + enc_data->fifo_byte_count;
+	int count = 0;
+	uint8_t sample_num = 0;
+
+	if ((uintptr_t)buffer_end <= *fit) {
+		return 0;
+	}
+
+	struct adc_data *data = (struct adc_data *)data_out;
+
+	memset(data, 0, sizeof(struct adc_data));
+	data->header.base_timestamp_ns = enc_data->timestamp;
+	data->header.reading_count = 1;
+
+	data->shift = adc_max32_convert_vref_to_shift(enc_data->vref_mv);
+
+	buffer += sizeof(struct adc_max32_fifo_data);
+	uint8_t sample_set_size = enc_data->sample_set_size;
+	/* Calculate which sample is decoded. */
+	if ((uint8_t *)*fit >= buffer) {
+		sample_num = ((uint8_t *)*fit - buffer) / sample_set_size;
+	}
+
+	while (count < max_count && buffer < buffer_end) {
+		const uint8_t *sample_end = buffer;
+
+		sample_end += sample_set_size;
+
+		if ((uintptr_t)buffer < *fit) {
+			/* This frame was already decoded, move on to the next frame */
+			buffer = sample_end;
+			continue;
+		}
+		/* 125 KSPS - this can be calculated from
+		 * cnt and idle dts parameters but it is hardcoded for now
+		 */
+		data->readings[count].timestamp_delta = sample_num * (UINT32_C(1000000000) / 62500);
+		adc_max32_convert_q31(&data->readings[count].value, buffer,
+				enc_data->max32_qscale_mode, enc_data->diff_mode,
+				enc_data->vref_mv, data->shift);
+
+		memcpy(out, buffer, 2);
+
+		buffer = sample_end;
+		*fit = (uintptr_t)sample_end;
```

Comment:
I would strongly suggest not using memory addresses for the frame iterator. Would data offset work for you? `fit` should be 0 on a first read and you can increment it by the number of bytes that you read in the while loop.